x-pack/plugins/security/server/routes/authentication/common.ts

```diff
@@ -71,4 +76,62 @@ export function defineCommonRoutes({ router, authc, basePath, logger }: RouteDef
       })
     );
   }
+
+  function getLoginAttemptForProviderType(providerType: string, redirectURL: string) {
+    const [redirectURLPath] = redirectURL.split('#');
+    const redirectURLFragment =
+      redirectURL.length > redirectURLPath.length
+        ? redirectURL.substring(redirectURLPath.length)
+        : '';
+
+    if (providerType === SAMLAuthenticationProvider.type) {
+      return { type: SAMLLogin.LoginInitiatedByUser, redirectURLPath, redirectURLFragment };
+    }
+
+    if (providerType === OIDCAuthenticationProvider.type) {
+      return { type: OIDCLogin.LoginInitiatedByUser, redirectURLPath };
+    }
+
+    return undefined;
+  }
+
+  router.post(
```

Comment:
__note:__ theoretically we can join `/login` and `login_with` endpoints now - I just didn't want to work on this in this PR. I'll file issue for that.